auth/src/services/aws/dynamodb.ts

```diff
@@ -0,0 +1,553 @@
+} from "@aws-sdk/client-dynamodb"
+import { marshall, unmarshall } from "@aws-sdk/util-dynamodb"
+import { DateTime } from 'luxon'
+import { didRegex } from '../../utils/did.js'
+import { now } from '../../utils/datetime.js'
+
+import {
+    ConfigKey,
+    createEmailVerificationCode,
+    Database,
+    DIDResult,
+    DIDStatus,
+    OTPStatus
+} from '../db.js'
+import { KeyService } from "../key.js"
+
+const AWS_REGION = process.env.AWS_REGION ?? ''
+const INITIAL_NONCE = '0'
+
+/**
+ * Returns current timestamp as unix integer
+ * @returns Unix timestamp
+ */
+const CONFIG_TABLE_NAME = process.env.IS_OFFLINE ? `cas-auth-config-${now()}` : (process.env.DB_CONFIG_TABLE_NAME ?? '')
+const DID_TABLE_NAME = process.env.IS_OFFLINE ? `cas-auth-did-${now()}` : (process.env.DB_DID_TABLE_NAME ?? '')
+const OTP_TABLE_NAME = process.env.IS_OFFLINE ? `cas-auth-otp-${now()}` : (process.env.DB_OTP_TABLE_NAME ?? '')
+
+class ItemNotFoundError extends Error { }
+
+export type Item = Record<string, AttributeValue>
+type ExpressionAttributeValues = Item
+
+export class DynamoDB implements Database {
+    name: string
+    readonly client: DynamoDBClient
+    readonly keyService: KeyService
+    private readonly _shouldCreateTableIfNotExists: boolean
+
+    constructor(createTableIfNotExists: boolean) {
+        this._shouldCreateTableIfNotExists = createTableIfNotExists
+        if (CONFIG_TABLE_NAME == '') {
+            throw Error('Missing DB_CONFIG_TABLE_NAME')
+        }
+        if (DID_TABLE_NAME == '') {
+            throw Error('Missing DB_DID_TABLE_NAME')
+        }
+        if (OTP_TABLE_NAME == '') {
+            throw Error('Missing DB_OTP_TABLE_NAME')
+        }
+        this.client = process.env.IS_OFFLINE
+            ? new DynamoDBClient({ endpoint: 'http://localhost:8000' })
+            : new DynamoDBClient({ region: AWS_REGION })
+    }
+
+    async init() {
+        for (const tableName of [CONFIG_TABLE_NAME, DID_TABLE_NAME, OTP_TABLE_NAME]) {
+            if (this._shouldCreateTableIfNotExists) {
+                await this._createTableIfNotExists(tableName)
+            } else {
+                const shouldThrow = true
+                await this._checkTableExists(tableName, shouldThrow)
+            }
+        }
+    }
+
+    private async _createTableIfNotExists(tableName: string): Promise<void> {
+        const shouldthrow = process.env.IS_OFFLINE ? false : true
+        if (await this._checkTableExists(tableName, shouldthrow)) return
+
+        const input: CreateTableCommandInput = {
+            TableName: tableName,
+            KeySchema: [
+                { AttributeName: 'PK', KeyType: 'HASH' },
+                { AttributeName: 'SK', KeyType: 'RANGE' }
+            ],
+            AttributeDefinitions: [
+                { AttributeName: 'PK', AttributeType: 'S' },
+                { AttributeName: 'SK', AttributeType: 'S' },
+                { AttributeName: 'GSI-1-PK', AttributeType: 'S' },
```

Comment:
I don't see the GSIs being used here, do we need them?

Reply:
They are added by default by serverless lift, so added the same schema here for consistency if we create a table outside of lift. Any harm in keeping them even if not used?

Reply:
I think that should be ok. In this case, since we're never writing to any fields that are part of the GSI, we probably won't be charged for their throughput.

When you get a chance, would you mind describing how the tables/indices are used (i.e. the read/write patterns)? Would to good to do a quick analysis to make sure there aren't any hot spots.

Reply:
Yes! will add to your calendar to discuss